### PR TITLE
load mode before creating master_params when fp16

### PR DESCRIPTION
**Pull request by wangdongxuking61**
When using fp16 and args.resume is True, the model should be loaded before creating master_params variable. Otherwise, the value in master_params is just random and not same as the loaded file because tensor in master_params is cloned before loaded model.

### CODE REVIEW

**Comment by mcarilli** [comment]
You're right, good catch.  However, looking at this more carefully, the best thing to do is save and restore the fp32 master params directly, and copy those into the model params when resuming.  I'll implement that myself.

**Comment by mcarilli** [comment]
Hopefully handled via https://github.com/NVIDIA/apex/commit/327b24469236e467ce3a91fac9fea6a3540af56d.  The examples now resume from checkpoints with the expected loss.

Let me know if this works for you.

**Comment by wangdongxuking61** [comment]
> Hopefully handled via [327b244](https://github.com/NVIDIA/apex/commit/327b24469236e467ce3a91fac9fea6a3540af56d). The examples now resume from checkpoints with the expected loss.
> 
> Let me know if this works for you.

well-done. I reviewed the code, it looks like right.
BTW, i think the following code 
```
saved_master_params = checkpoint['master_params']
for master, saved in zip(master_params, saved_master_params):
    master.data.copy_(saved.data)
```
can be replaced by 
```
if args.fp16:
    saved_master_params = checkpoint['master_params']
    for master, saved in zip(master_params, saved_master_params):
        master.data.copy_(saved.data)
```
because when using fp32, `master_params` is just reference to model's Parameters, calling `model.load_state_dict(checkpoint['state_dict'])` is enough.
